server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,180 @@
+                        builder.startObject(SYNONYMS_SET_FIELD);
+                        {
+                            builder.field("type", "keyword");
+                        }
+                        builder.endObject();
+                    }
+                    builder.endObject();
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to build mappings for " + SYNONYMS_INDEX_CONCRETE_NAME, e);
+        }
+    }
+
+    public void putSynonymsset(String resourceName, SynonymsSet synonymsset, ActionListener<PutSynonymsAction.Response> listener) {
+
+        // TODO Add synonym rules validation
+
+        // Delete synonyms set if it existed previously. Avoid catching an index not found error by ignoring unavailable indices
+        DeleteByQueryRequest dbqRequest = new DeleteByQueryRequest(SYNONYMS_ALIAS_NAME).setQuery(
+            QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName)
+        ).setIndicesOptions(IndicesOptions.fromOptions(true, true, false, false));
+
+        client.execute(
+            DeleteByQueryAction.INSTANCE,
+            dbqRequest,
+            listener.delegateFailure((deleteByQueryResponseListener, bulkByScrollResponse) -> {
+                boolean created = bulkByScrollResponse.getDeleted() == 0;
+                final List<BulkItemResponse.Failure> bulkFailures = bulkByScrollResponse.getBulkFailures();
+                if (bulkFailures.isEmpty() == false) {
+                    listener.onFailure(
+                        new ElasticsearchException(
+                            "Error updating synonyms: "
+                                + bulkFailures.stream().map(BulkItemResponse.Failure::getMessage).collect(Collectors.joining("\n"))
+                        )
+                    );
+                }
+
+                // Insert as bulk requests
+                BulkRequestBuilder bulkRequestBuilder = client.prepareBulk();
+                try {
+                    for (SynonymRule synonymRule : synonymsset.synonyms()) {
+
+                        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                            builder.startObject();
+                            {
+                                builder.field(SYNONYMS_FIELD, synonymRule.synonym());
+                                builder.field(SYNONYMS_SET_FIELD, resourceName);
+                            }
+                            builder.endObject();
+
+                            final IndexRequest indexRequest = new IndexRequest(SYNONYMS_ALIAS_NAME).opType(DocWriteRequest.OpType.INDEX)
+                                .source(builder);
+                            final String synonymRuleId = synonymRule.id();
+                            if (synonymRuleId != null) {
+                                indexRequest.id(synonymRuleId);
```

Comment:
If I understand this correctly, using the plain synonym rule id a doc id will potentially overwrite rules using the same id in other synonym sets? I think joining set id / rule Id might be useful as a doc id here.
In that case I also believe the previous delete request isn't necessary because each rule will overwrite any potential predecessor?

Reply:
> If I understand this correctly, using the plain synonym rule id a doc id will potentially overwrite rules using the same id in other synonym sets?

That was not the plan - rule set IDs would be isolated from one another. Potentially, two rule sets could contain the same rule ID but the two would be independent.

That's why `ruleset_id` is stored as a separate field in the rule docs - it is used as a filter to retrieve all the rules in a ruleset.

> In that case I also believe the previous delete request isn't necessary because each rule will overwrite any potential predecessor?

The delete request would still be needed, as:
- Users can choose not to specify rule IDs (because they don't care about them, or are not planning to individually update rules)
- Users may "delete" rules by replacing a ruleset, but not specifying some of the rules that were part of the previous one.

Reply:
You're right Christoph, thanks for explaining. I was taking advantage of using the ID auto-generation feature from the index request, and didn't thought on potential collisions with user-defined IDs. 

I like your approach to ensure that IDs are unique to each synonym set, by prepending the rule set ID to it.

In case the user does not specify an ID, we could generate one using `synonymSetId + "_" + UUIDs.base64UUID()`. 
 Do you think this approach would work?

Reply:
Yes, as long as the UI can safely remove that prefix when displaying the ids. 
Is the idea to show auto-generated Ids in error messages and when listing all rules of a set then? Or leave those empty in that case?

Reply:
We still haven't thought about that on the UI side. I'd say that ids should not be displayed, as they do not convey useful information to the user (who is using the UI to manage synonyms).

IDs make sense to me from the API perspective, but not from the UI perspective. That is the way synonyms work for App Search today - auto-generated IDs are used for synonyms, but only used on the API side.